Set first_published_at from edition history

steventux opened the pull request:
https://trello.com/c/BHzWqzPF/1043-3-sort-out-the-same-firstpublishedat-date-for-1-4-of-the-database

We have 800000 records with a `first_published_at` on 2016-02-29,
a quarter of these are placeholders, just over 200000 have a
'First published' change history item, so use the date from this
to set the correct `first_published_at` value.

kevindew commented on db/migrate/20170825144733_repopulate_first_published_at_from_change_history.rb:
```diff
@@ -0,0 +1,39 @@
+
+  def up
+    min_date = '2016-02-29'
+    max_date = '2016-02-29T23:59:59'
```
I'm not sure you have to be dealing with a 24 hour window here. I think it's all the same second.

```
[3] pry(main)> Edition.where("first_published_at > ? AND first_published_at < ?", "2016-02-29", "2016-02-29T23:59:59").count
   (1055.6ms)  SELECT COUNT(*) FROM "editions" WHERE (first_published_at > '2016-02-29' AND first_published_at < '2016-02-29T23:59:59')
=> 789244
```

```
[12] pry(main)> Edition.where("first_published_at > ? AND first_published_at <= ?", "2016-02-29T09:24:09", "2016-02-29T09:24:11").count
   (1056.2ms)  SELECT COUNT(*) FROM "editions" WHERE (first_published_at > '2016-02-29T09:24:09' AND first_published_at <= '2016-02-29T09:24:11')
=> 788272
```

You might know better than me at this point though whether the whole day is a problem or just that second.

steventux replied:
I think you're right, somewhere in analysing this I got the idea there were 2 different bulk update times that day, evidently not though:
```
SELECT COUNT(*) AS count_all, "editions"."first_published_at" AS editions_first_published_at FROM "editions" WHERE (first_published_at BETWEEN '2016-02-29' AND '2016-02-29T23:59:59') GROUP BY "editions"."first_published_at" HAVING (count(*) > 1000)
=> {Mon, 29 Feb 2016 09:24:10 UTC +00:00=>809051}
```

thomasleese commented on db/migrate/20170825144733_repopulate_first_published_at_from_change_history.rb:
```diff
@@ -0,0 +1,39 @@
+
+    # ~395 seconds
+    results.each do |edition_id, document_id, timestamp|
+      sql = "UPDATE editions SET first_published_at = '#{timestamp}' WHERE id = #{edition_id};"
```
If the slowness here is problematic, I wonder if we can make this faster by grouping lots of queries together with semi-colons into batches and then calling `execute` only a few times.

steventux replied:
:+1: I'll batch them up, this migration isn't blocking but I'm sure it can be run quicker than 6 minutes.

steventux commented:
@thomasleese It was just over 2 minutes quicker, the initial find is slow because of the json column lookups. I've rebased this (and moved the migration timestamp on because of conflicts) so could you re-review?
